src/plugins/intel_cpu/src/nodes/matmul.cpp

```diff
@@ -110,8 +110,75 @@ bool MatMul::isSupportedOperation(const std::shared_ptr<const ngraph::Node>& op,
     return true;
 }
 
+class MMShapeInfer : public ShapeInferEmptyPads {
+public:
+    MMShapeInfer(const std::shared_ptr<const ngraph::opset1::MatMul>& op) : m_matmul(op) {
+        m_out_rank = m_matmul->get_output_partial_shape(0).rank().get_length();
+        m_YShape = VectorDims(m_out_rank, 1); // for output and cache
+    }
+    std::vector<VectorDims> infer(
+        const std::vector<std::reference_wrapper<const VectorDims>>& input_shapes,
+        const std::unordered_map<size_t, MemoryPtr>& data_dependency) override {
+        VectorDims AShape = input_shapes[0].get();
+        VectorDims BShape = input_shapes[1].get();
+        const size_t& Aranks = AShape.size();
+        const size_t& Branks = BShape.size();
+
+        // getSupportedDescriptors has done some shape check.
+        // 1. Needn't assert the scalar type since the matmul_shape_inference has checked.
+        // 2. Needn't check the compatibility of the last two dims
+        // 3. 1-D x 1-D is needed
+        // 4. transpose is necessary
+        // 5. Just support the same rank of matmul
+        // 6. simplify the broadcast check
+        if (Aranks == 1 && Branks == 1 && AShape[0] == BShape[0]) {
+            return {m_YShape};
+        }
+
+        // transpose
+        const bool& transpose_a = m_matmul->get_transpose_a();
+        const bool& transpose_b = m_matmul->get_transpose_b();
+
+        m_YShape.at(m_out_rank-2) = transpose_a ? AShape[Aranks-1] : AShape[Aranks-2];
+        m_YShape.at(m_out_rank-1) = transpose_b ? BShape[Branks-2] : BShape[Branks-1];
+
+        for (size_t i=0; i < m_out_rank-2; ++i) {
+            size_t max = std::max(AShape.at(i), BShape.at(i));
+            size_t min = std::min(AShape.at(i), BShape.at(i));
+            if ((AShape.at(i) == BShape.at(i)) || (min == 1)) {
+                m_YShape[i] = max;
+            } else {
+                IE_THROW() << "Incompatible MatMul batch dimension. Cant merge the first input dimension=" <<
+                              AShape[i] << " with second input dimension=" << BShape[i] << " at index=" << i;
+            }
+        }
+
+        return {m_YShape};
+    }
+
+    port_mask_t get_port_mask() const override {
+        return EMPTY_PORT_MASK;
+    }
+
+private:
+    const std::shared_ptr<const ngraph::opset1::MatMul> m_matmul;
```

Comment:
I would say it is better to store `bool transpose_a` and `bool transpose_b`. So it gives faster access to those fields than through the operation pointer.

Reply:
> I would say it is better to store `bool transpose_a` and `bool transpose_b`. So it gives faster access to those fields than through the operation pointer.

Yes. move it to private member.

Reply:
> Why do we still have this data member then?

Sry for this. Have removed it.

Reply:
> Why do we still have this data member then?

Sry for this. Have removed it. thx